### PR TITLE
build: Optimize docker images

### PR DESCRIPTION
**Pull request by kesara**
As per comments in https://github.com/ietf-tools/xml2rfc/pull/849#issuecomment-1210075932, attempt to optimize the docker images generated by xml2rfc.

### CODE REVIEW

**Comment by kesara** [comment]
@martinthomson I can't add you as a reviewer because of the project restrictions to stop spam, but appreciate your comments on this PR.

File: docker/base.Dockerfile [hunk 1]
```diff
@@ -2,75 +2,60 @@ FROM ubuntu:jammy
...
+RUN make install && \
+    apt-get remove --purge -y software-properties-common wget unzip && \
+    apt-get autoclean && \
+    apt-get clean && \
+    pip3 uninstall -y decorator dict2xml pypdf2 && \
+    rm setup.py Makefile configtest.py requirements.txt && \
+    rm -r xml2rfc build dist && \
+    echo 'export LANG=en_US.UTF-8' >> ~/.bashrc && \
+    echo 'xml2rfc --version' >> ~/.bashrc && \
+    echo 'if [ -d ~/xml2rfc ]; then cd ~/xml2rfc; fi' >> ~/.bashrc
```
Comment (NGPixel) [inline docker/base.Dockerfile:61]: Should be a separate RUN statement, as it's not related to what's above.

File: docker/dev.Dockerfile [hunk 2]
```diff
@@ -2,46 +2,32 @@ FROM ghcr.io/ietf-tools/xml2rfc-base:latest
...
-
-# cleanup
-RUN apt-get remove --purge -y software-properties-common
-RUN apt-get autoclean
-RUN apt-get clean
-
-# bash config
-RUN echo 'xml2rfc --version --verbose' >> ~/.bashrc
+    "pypdf2>=2.6.0" && \
+    echo 'xml2rfc --version --verbose' >> ~/.bashrc
```
Comment (NGPixel) [inline docker/dev.Dockerfile:31]: Should be a separate RUN statement, like above.

**Review by NGPixel — CHANGES_REQUESTED** [review]

File: docker/base.Dockerfile [hunk 3]
```diff
@@ -2,75 +2,60 @@ FROM ubuntu:jammy
...
+        libxml2-utils \
+        groff \
+        wget \
+        unzip \
+        locales \
+        python3.10 \
+        python3.10-dev \
+        python3-pip \
+        python3.10-distutils && \
+    locale-gen en_US.UTF-8
```
Comment (martinthomson) [inline docker/base.Dockerfile:25]: To remove cached and temporary files, add:
```
 && rm -rf /var/lib/apt/lists/* /var/log/dpkg.log \
 && apt-get autoremove -y && apt-get clean -y
```

File: docker/base.Dockerfile (hunk 1 above)
Comment (martinthomson) [reply docker/base.Dockerfile:61]: Yeah, and you could move the last three lines further up as well (they are not dependent on anything that happens at this later stage.  (Before the COPY line even.)

**Review by martinthomson — COMMENTED** [review]
A few tips, but this should be a lot faster and leaner as-is.

File: docker/base.Dockerfile [hunk 4]
```diff
@@ -2,75 +2,60 @@ FROM ubuntu:jammy
...
-RUN echo 'if [ -d ~/xml2rfc ]; then cd ~/xml2rfc; fi' >> ~/.bashrc
+# Build xml2rfc & finalize
+RUN make install && \
+    apt-get remove --purge -y software-properties-common wget unzip && \
+    apt-get autoclean && \
+    apt-get clean && \
+    pip3 uninstall -y decorator dict2xml pypdf2 && \
+    rm setup.py Makefile configtest.py requirements.txt && \
+    rm -r xml2rfc build dist && \
+    echo 'export LANG=en_US.UTF-8' >> ~/.bashrc && \
```
Comment (martinthomson) [inline docker/base.Dockerfile:59]: This might be an ENV instruction.

File: docker/base.Dockerfile [hunk 5]
```diff
@@ -2,75 +2,60 @@ FROM ubuntu:jammy
...
-RUN rm setup.py Makefile configtest.py requirements.txt
-RUN rm -r xml2rfc build dist
-
-# bash config
-RUN echo 'export LANG=en_US.UTF-8' >> ~/.bashrc
-RUN echo 'xml2rfc --version' >> ~/.bashrc
-RUN echo 'if [ -d ~/xml2rfc ]; then cd ~/xml2rfc; fi' >> ~/.bashrc
+# Build xml2rfc & finalize
+RUN make install && \
+    apt-get remove --purge -y software-properties-common wget unzip && \
```
Comment (martinthomson) [inline docker/base.Dockerfile:53]: Removing things doesn't really help save on space.  It increases image size and build time, contrary to what you might expect.  Besides, these seem like useful tools to have :)

The way to build a software package from source without inflating image size is a little counter-intuitive.  Do that by creating two images - a big one and a small one - and copying the build product from the big image to the small one.  See https://github.com/mozilla/neqo/blob/main/qns/Dockerfile

**Review by martinthomson — COMMENTED** [review]
Now for the fun comments...

It really depends on how far down the rabbit hole you want to get.

File: docker/base.Dockerfile (hunk 1 above)
Comment (kesara) [reply docker/base.Dockerfile]: I've moved these statements to the top.

File: docker/dev.Dockerfile (hunk 2 above)
Comment (kesara) [reply docker/dev.Dockerfile]: This line wasn't necessary, so I remove it.

File: docker/base.Dockerfile (hunk 3 above)
Comment (kesara) [reply docker/base.Dockerfile]: I've added these commands to both base and dev Dockerfiles.

File: docker/base.Dockerfile (hunk 5 above)
Comment (kesara) [reply docker/base.Dockerfile]: I don't want to move to that system yet, so I've removed the `apt-get remove` command.

Comment (martinthomson) [reply docker/base.Dockerfile]: Yeah, it's a fairly big lift and probably overkill for what you need here.

**Review by NGPixel — APPROVED** [review]

**Pull request merged**